common/src/Exceptions.h

```diff
@@ -92,6 +92,11 @@ namespace TrenchBroom {
         using Exception::Exception;
     };
 
+    class NodeCloneException : public Exception {
```

Comment:
Why can cloning fail? I don't think we should model it as being something that can fail.

The implementation of Brush::clone will need to be adjusted, because we're currently using a method that can fail (even though it really shouldn't, since the starting brush is presumably valid). It should instead use the copy constructor on Polyhedron, clone the BrushFaces, and make sure the face pointers are linked up correctly.

BTW, what is the situation now as far as instances of Brush* or Node* being in a valid state? We used to have the situation where if Brush::clip threw an exception, you were left with a "broken" Brush* object which was the caller's responsibility to destroy and not use. I assume this is changed and a Brush* is always valid now?

Reply:
> The implementation of Brush::clone will need to be adjusted, because we're currently using a method that can fail (even though it really shouldn't, since the starting brush is presumably valid). It should instead use the copy constructor on Polyhedron, clone the BrushFaces, and make sure the face pointers are linked up correctly.

Ok, I'll look into that.

> BTW, what is the situation now as far as instances of Brush* or Node* being in a valid state? We used to have the situation where if Brush::clip threw an exception, you were left with a "broken" Brush* object which was the caller's responsibility to destroy and not use. I assume this is changed and a Brush* is always valid now?

I haven't changed anything about that - but we should fix it. The easiest way would be to perform the operation on a copy, then move the copy into the object being manipulated (or swapping them).

Reply:
Ok, I have dropped the commit that changed Node::clone entirely, and have replaced it with a commit that makes Brush::doClone safe. Nice!

Reply:
Brush now has copy, move and swap, which will come in handy when we go about giving it a stronger guarantee in case of failure, that is, the an operation should either succeed, or not leave the brush / entity / whatever in its original state.